src/analyzer.ts

```diff
@@ -73,45 +113,88 @@ export class Analyzer {
 
     // Note we can't use Promise.all here, because new promises can be added to
     // the promises array as long as any promise is pending.
-    const errors = [];
-    while (this.#placeholderUpgradePromises.length > 0) {
-      const result = await this.#placeholderUpgradePromises.shift();
-      if (result?.ok === false) {
-        errors.push(result.error);
+    await this.#waitForanalysisToComplete();
+    {
+      const errors = await this.getDiagnostics();
+      if (errors.size > 0) {
+        return {ok: false, error: [...errors]};
       }
     }
-    if (errors.length > 0) {
-      return {ok: false, error: errors};
-    }
 
     // We can safely assume all placeholders have now been upgraded to full
     // configs.
-    const rootConfig = rootPlaceholder as ScriptConfig;
+    const rootConfig = rootPlaceholder.placeholder;
+    if (rootConfig.state === 'unvalidated') {
+      throw new Error(
+        `Internal error: script ${root.name} in ${root.packageDir} is still unvalidated but had no failures`
+      );
+    }
     const cycleResult = this.#checkForCyclesAndSortDependencies(
       rootConfig,
       new Set()
     );
     if (!cycleResult.ok) {
       return {ok: false, error: [cycleResult.error]};
     }
-    return {ok: true, value: rootConfig};
+    const validRootConfig = cycleResult.value;
+    return {ok: true, value: validRootConfig};
+  }
+
+  async getDiagnostics(): Promise<Set<Failure>> {
```

Comment:
Can this be `#private`?

Reply:
I think it might need to be later on but we can make it private for now.